packages/kbn-std/src/is_internal_url.ts

```diff
@@ -8,35 +8,63 @@
 
 import { parse as parseUrl } from 'url';
 
+// Used as the dummy base URL for WHATWG URL parser only. We use custom `estc` protocol since it allows us to use empty
+// host: https://url.spec.whatwg.org/#empty-host.
+export const CUSTOM_BASE_URL = 'estc://';
+
+// The regex to match characters that are automatically removed from the URL by WHATWG URL parser:
+// https://infra.spec.whatwg.org/#ascii-tab-or-newline
+const TAB_AND_NEW_LINE_REGEX = /[\r\n\t]/g;
+
 /**
  * Determine if url is outside of this Kibana install.
  */
 export function isInternalURL(url: string, basePath = '') {
+  // First, perform a basic check for the absolute URLs since only such URLs can be parsed without base URL. Also, if
+  // the URL cannot be parsed with a custom base URL, then it's not a valid URL, and we should also return `false`.
+  // @ts-expect-error `canParse` isn't defined in types yet.
+  if (URL.canParse(url) || !URL.canParse(url, CUSTOM_BASE_URL)) {
+    return false;
+  }
+
+  // Now, let's use the WHATWG URL parser, the one used by browsers, to parse the URL. Since WHATWG URL parser doesn't
+  // support parsing relative URLs as is, we have to use a custom base URL with an empty host. If the parsed URL has a
+  // non-empty host, then the provided URL is either an absolute or scheme-relative (//) URL.
+  const normalizedURL = new URL(url, CUSTOM_BASE_URL);
+  if (normalizedURL.host) {
+    return false;
+  }
+
+  // The fact that we have to use a custom base URL with an empty host doesn't allow us to distinguish proper relative
```

Comment:
__note:__ alternatively, we can just use `URL` parser twice. The second time would be with something like `estc://host` and if `host` of the parsed URL is still empty string then we can say that the specific URL has an empty host. Maybe it'd even be better. What would you prefer?